src/error.rs

```diff
@@ -97,29 +97,166 @@ impl From<OddLengthStringError> for ToBytesError {
     fn from(e: OddLengthStringError) -> Self { Self::OddLengthString(e) }
 }
 
-/// Invalid hex character.
+/// UTF-8 character is not an ASCII hex character.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct InvalidCharError {
-    pub(crate) invalid: u8,
+    pub(crate) invalid: char,
     pub(crate) pos: usize,
 }
 
 impl InvalidCharError {
-    /// Returns the invalid character byte.
-    pub fn invalid_char(&self) -> u8 { self.invalid }
-    /// Returns the position of the invalid character byte.
+    /// Returns the invalid character.
+    pub fn invalid_char(&self) -> char { self.invalid }
+
+    /// Returns the position in the input string of the invalid character.
     pub fn pos(&self) -> usize { self.pos }
 }
 
 impl fmt::Display for InvalidCharError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "invalid hex char {} at pos {}", self.invalid_char(), self.pos())
+        write!(
+            f,
+            "char {} at pos {} is not a valid ASCII hex character",
+            self.invalid_char(),
+            self.pos()
+        )
     }
 }
 
 #[cfg(feature = "std")]
 impl std::error::Error for InvalidCharError {}
 
+/// Invalid ASCII encoding of a hex digit.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InvalidDigitError {
+    /// The high digit in the pair e.g., for 'ab' this is 'a'.
+    pub(crate) hi: u8,
+
+    /// The high digit in the pair e.g., for 'ab' this is 'b'.
+    pub(crate) lo: u8,
+
+    /// `true` if the invalid byte is the `hi` one, `false` if its the `lo` one.
+    pub(crate) is_hi: bool,
+
+    /// Position of the stat of the byte pair in the input string.
+    pub(crate) pos: usize,
+}
+
+impl InvalidDigitError {
+    /// Returns the invalid byte.
+    pub fn invalid_byte(&self) -> u8 {
+        if self.is_hi {
+            self.hi
+        } else {
+            self.lo
+        }
+    }
+
+    /// Returns the character position in input string of the invalid byte.
+    pub fn pos(&self) -> usize {
+        if self.is_hi {
+            self.pos
+        } else {
+            self.pos + 1
+        }
+    }
+
+    pub(crate) fn into_invalid_char_error(
+        self,
+        // The next item from the iterator.
+        next: Option<Result<u8, InvalidDigitError>>,
+    ) -> InvalidCharError {
+        // Try to map the UTF-8 values in `v` to an invalid character.
+        fn error(v: &[u8], pos: usize) -> Option<InvalidCharError> {
+            core::str::from_utf8(v)
+                .map(|s| InvalidCharError { invalid: s.chars().next().unwrap(), pos })
```

Comment:
In ce28e9ba01cda818439b36108e072435715552fc:

It seems like this function will return `None` in the case that you give it bad UTF-8 (for example, a non-ASCII character immediately followed by an ASCII character, which never happens in UTF-8 because it would make it ambiguous how to scan backward from the ASCII character to the nearest character boundary.)

But then below you seem to interpret `None` as "promote one of the bad digits in `self` to a `char`". Which, aside from being logically wrong, might not even result in a valid character.